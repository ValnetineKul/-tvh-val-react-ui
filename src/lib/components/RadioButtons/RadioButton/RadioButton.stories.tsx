import { useArgs } from '@storybook/client-api';
import type { ComponentProps } from 'react';
import React from 'react';
import type { StoryTemplate } from '../../../story-utils';
import { createMeta } from '../../../story-utils';
import RadioButton from './RadioButton';

type Props = ComponentProps<typeof RadioButton>;
type Args = Omit<Props, 'onChange'>;

const Template = ((args: Args) => {
  const [{ checked }, updateArgs] = useArgs();

  return (
    <RadioButton
      {...args}
      checked={checked}
      onChange={(...params) => {
        updateArgs({ checked: params[1] });
      }}
    />
  );
}) as StoryTemplate<Args>;

export const Primitive = Template.bind({});
Primitive.args = {
  label: 'RadioButton label',
  value: 'RadioButton value',
  name: 'radio-button',
  checked: false,
  disabled: false,
};

Primitive.storyName = 'RadioButton';

export default createMeta({
  component: RadioButton,
  title: 'Components/RadioButtons/RadioButton',
  argTypes: {
    onChange: {
      action: { argTypesRegex: '^on.*' },
      table: {
        type: {
          detail:
            'Callback fired when the state is changed: (event: React.ChangeEvent<HTMLInputElement>, checked: boolean) => void.',
        },
      },
    },
    value: {
      table: {
        type: {
          detail: 'Value of the selected radio button.',
        },
      },
    },
  },
});
