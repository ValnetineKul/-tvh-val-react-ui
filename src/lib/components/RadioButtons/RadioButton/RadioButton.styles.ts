import { makeStyles } from '../../../../themes/core';

const useStyles = makeStyles<void, 'disabled' | 'label'>({ name: 'RadioButton' })((theme, _, classes) => ({
  root: {
    alignItems: 'flex-start',
    marginLeft: theme.layoutSpacing['spacing/000'],
    marginRight: theme.layoutSpacing['spacing/000'],
    [`&.${classes.disabled} .${classes.label}`]: {
      color: theme.color['text/onSurfaceCurrent/disabled'],
    },
  },
  radio: {
    padding: theme.layoutSpacing['spacing/000'],
    margin: theme.layoutSpacing['spacing/100'],
    backgroundColor: theme.color['bg/surface100/default'],
    '&:hover': {
      backgroundColor: theme.color['bg/surface100/default'],
    },
    '&:focus-within': {
      boxShadow: `0px 0px 0px 4px ${theme.color['border/onSurfaceCurrent/focus']}`,
    },
  },
  disabled: {},
  label: {
    marginLeft: theme.layoutSpacing['spacing/300'],
  },
  icon: {
    width: `calc( ${theme.layoutSpacing['spacing/400']} + 4px)`,
    height: `calc( ${theme.layoutSpacing['spacing/400']} + 4px)`,
    borderRadius: `calc(${theme.borderRadius['borderRadius/circle']} / 2)`,
  },
  unchecked: {
    border: `2px solid ${theme.color['border/onSurfaceCurrent/default']}`,
  },
  disabledUnchecked: {
    borderColor: theme.color['border/onSurfaceCurrent/disabled'],
    backgroundColor: theme.color['bg/disabled'],
  },
  checkedIcon: {
    boxShadow: `inset 0px 0px 0px 2px ${theme.color['bg/surfacePrimary/default']}, inset 0px 0px 0px 5px ${theme.color['bg/surface100/default']}, inset 0px 0px 0px 10px ${theme.color['bg/surfacePrimary/default']}`,
  },
  disabledCheckedIcon: {
    boxShadow: `inset 0px 0px 0px 2px ${theme.color['bg/disabled']}, inset 0px 0px 0px 5px ${theme.color['bg/surface100/default']}, inset 0px 0px 0px 10px ${theme.color['bg/disabled']}`,
  },
}));

export default useStyles;
