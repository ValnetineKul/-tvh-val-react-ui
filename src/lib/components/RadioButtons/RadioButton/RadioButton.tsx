import type { FC } from 'react';
import React from 'react';
import FormControlLabel from '@mui/material/FormControlLabel';
import { default as MuiRadio } from '@mui/material/Radio';

import Typography from '../../Typography';
import useStyles from './RadioButton.styles';
import type { DataAttributes } from '../../../types/common';

export interface RadioButtonProps {
  label: string;
  value: string;
  name?: string;
  checked?: boolean;
  disabled?: boolean;
  className?: string;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>, checked: boolean) => void;
  inputProps?: React.HTMLAttributes<HTMLInputElement> & DataAttributes;
}

const RadioButton: FC<RadioButtonProps> = ({
  label,
  value,
  name,
  checked,
  disabled = false,
  className,
  onChange,
  inputProps,
}) => {
  const { classes, cx } = useStyles();

  const renderIcon = () => {
    return (
      <span
        className={cx(classes.icon, {
          [classes.disabledUnchecked]: disabled,
          [classes.unchecked]: !disabled,
        })}
      />
    );
  };

  const renderCheckedIcon = () => {
    return (
      <span
        className={cx(classes.icon, {
          [classes.disabledCheckedIcon]: disabled,
          [classes.checkedIcon]: !disabled,
        })}
      />
    );
  };

  const renderRadio = () => {
    return (
      <MuiRadio
        inputProps={inputProps}
        name={name}
        value={value}
        className={classes.radio}
        checked={checked}
        disabled={disabled}
        onChange={onChange}
        icon={renderIcon()}
        checkedIcon={renderCheckedIcon()}
        disableRipple
      />
    );
  };

  if (!label) {
    return renderRadio();
  }

  return (
    <FormControlLabel
      classes={{ root: classes.root, label: cx(classes.label, className), disabled: classes.disabled }}
      label={<Typography component="span">{label}</Typography>}
      value={value}
      control={renderRadio()}
    />
  );
};

export default RadioButton;
