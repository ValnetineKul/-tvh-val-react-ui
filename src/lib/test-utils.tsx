import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
