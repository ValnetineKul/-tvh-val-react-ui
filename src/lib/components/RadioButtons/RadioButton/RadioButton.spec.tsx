import React from 'react';
import { userEvent, render } from '../../../test-utils';
import RadioButton from './RadioButton';

const props = {
  label: 'RadioButton label',
  value: 'RadioButton value',
};

describe('RadioButton', () => {
  it('Should render correctly', () => {
    const { container } = render(<RadioButton {...props} onChange={() => null} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should disable radio button if disabled prop is true', () => {
    const { getByRole } = render(<RadioButton {...props} onChange={() => null} disabled />);
    const radioButton = getByRole('radio') as HTMLInputElement;
    expect(radioButton).toBeDisabled();
  });

  it('Should have radio button checked if checked prop is true', () => {
    const { getByRole } = render(<RadioButton {...props} onChange={() => null} checked />);
    const radioButton = getByRole('radio') as HTMLInputElement;
    expect(radioButton).toBeChecked();
  });

  it('Should trigger onChange', () => {
    const onChange = jest.fn();
    const { getByRole } = render(<RadioButton {...props} onChange={onChange} />);
    const radioButton = getByRole('radio') as HTMLInputElement;
    userEvent.click(radioButton);
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it('Should show text label', () => {
    const { getByLabelText } = render(<RadioButton {...props} onChange={() => null} />);
    expect(getByLabelText('RadioButton label', { selector: 'input' })).toBeInTheDocument();
  });

  it('Should not have label tag if label prop is empty', () => {
    const { queryByLabelText } = render(<RadioButton {...props} label="" onChange={() => null} />);
    expect(queryByLabelText('RadioButton label', { selector: 'input' })).not.toBeInTheDocument();
  });

  it('Should have "value" attribute if label prop is empty', () => {
    const { getByRole } = render(<RadioButton {...props} label="" onChange={() => null} />);
    const radioButton = getByRole('radio') as HTMLInputElement;
    expect(radioButton).toHaveAttribute('value', 'RadioButton value');
  });

  it('Should set name attribute', () => {
    const { getByRole } = render(<RadioButton {...props} name="radio-button" onChange={() => null} />);
    const radioButton = getByRole('radio') as HTMLInputElement;
    expect(radioButton).toHaveAttribute('name', 'radio-button');
  });
});
