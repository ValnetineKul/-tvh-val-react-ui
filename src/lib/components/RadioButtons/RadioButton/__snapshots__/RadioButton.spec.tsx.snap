// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton Should render correctly 1`] = `
<label
  class="tss-1neb9gh-RadioButton-root MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd mui-j204z7-MuiFormControlLabel-root"
>
  <span
    class="MuiRadio-root MuiRadio-colorPrimary MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root tss-1tmse2r-RadioButton-radio mui-vqmohf-MuiButtonBase-root-MuiRadio-root"
  >
    <input
      class="PrivateSwitchBase-input mui-1m9pwf3"
      type="radio"
      value="RadioButton value"
    />
    <span
      class="tss-1r3d2cc-RadioButton-icon-RadioButton-unchecked"
      font-size="medium"
    />
  </span>
  <span
    class="MuiTypography-root MuiTypography-body1 tss-17hy40x-RadioButton-label tss-1-RadioButton-label-ref MuiFormControlLabel-label mui-b600p5-MuiTypography-root"
  >
    <span
      class="tss-0-Typography-weightRegular-ref tss-6tt2d9-Typography-root-Typography-weightRegular-Typography-body500"
    >
      RadioButton label
    </span>
  </span>
</label>
`;
